specs/core/0_beacon-chain.md

```diff
@@ -1352,22 +1353,20 @@ def initiate_validator_exit(state: BeaconState, index: ValidatorIndex) -> None:
     Note that this function mutates ``state``.
     """
     validator = state.validator_registry[index]
-    validator.initiated_exit = True
-```
-
-#### `exit_validator`
-
-```python
-def exit_validator(state: BeaconState, index: ValidatorIndex) -> None:
-    """
-    Exit the validator with the given ``index``.
-    Note that this function mutates ``state``.
-    """
-    validator = state.validator_registry[index]
-
-    # Update validator exit epoch if not previously exited
+    # Operation is a no-op if validator is already in the queue
     if validator.exit_epoch == FAR_FUTURE_EPOCH:
-        validator.exit_epoch = get_delayed_activation_exit_epoch(get_current_epoch(state))
+        # Update exit queue counters
+        if state.exit_epoch < get_delayed_activation_exit_epoch(get_current_epoch(state)):
+            state.exit_epoch = get_delayed_activation_exit_epoch(get_current_epoch(state))
+        if state.exit_queue_filled >= MAX_EXIT_DEQUEUES_PER_EPOCH:
+            state.exit_epoch += 1
+            state.exit_queue_filled = 0
+        # Set validator exit epoch and withdrawable epoch
+        if validator.exit_epoch > state.exit_epoch:
+            validator.exit_epoch = state.exit_epoch
```

Comment:
Should we set `exit_epoch` to something even if `validator.exit_epoch <= state.exit_epoch`?

Reply:
Why? That implies that the validator is already in the exit queue.

Reply:
If `initiate_validator_exit` is called (e.g. because `slash_validator` is called) then we want that event to be recorded somehow in the `Validator` object. Do we need `validator.initiated_exit` or `validator.exitable_epoch`?

Reply:
My thinking is that you can check that a validator has initiated the exit process via `validator.exit_epoch != FAR_FUTURE_EPOCH`.

Reply:
I think what was tripping me up is the condition `validator.exit_epoch > state.exit_epoch`. It seems that this condition is always true since it is inside `if validator.exit_epoch == FAR_FUTURE_EPOCH`, and we always have `FAR_FUTURE_EPOCH > state.exit_epoch`.

Reply:
Ah, you're right there! fixed.